keps/sig-scheduling/20200116-coscheduling-plugin-based-on-scheduler-framework.md

```diff
@@ -92,51 +92,51 @@ type Coscheduling struct {
+
+Firstly, we will inherit the default in-tree PrioritySort plugin so as to honor .spec.priority to ensure high-priority Pods are always sorted ahead of low-priority ones.
+
+Secondly, if two Pods hold the same priority, the sorting precedence is described as below:
+
+- If they are both regularPods (without particular PodGroup label), compare their `InitialAttemptTimestamp` field: the Pod with earlier `InitialAttemptTimestamp` is positioned ahead of the other.
+  
+- If one is regularPod and the other is pgPod, compare regularPod's `InitialAttemptTimestamp` with pgPod's `InitialTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
+  
+- If they are both pgPods:
+  - Compare their `InitialAttemptTimestamp`: the Pod with earlier timestamp is positioned ahead of the other.
+  - If their `InitialAttemptTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `InitialAttemptTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)
+
+**Note1**: There are different `InitialTimestamp` (even if they are the same, the UID will be different). So when the pods enter the queue, the pods that belongs to the same PodGroup will be together.
 
 #### Pre-Filter
-1. `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.
+When a `pgGroup` Pod is being scheduled for the first time, we have 2 option to deal with it: either start the full scheduling cycle no matter its grouping Pods are present inside schedule queue, or fail quick its scheduling cycle as its grouping Pods number doesn't meet `minAvailable`. The former case is more efficient, but may cause partial Pods holding system resources until a timeout. The latter case may result in extra scheduling cycles (even if we're going to fail them fast), but will reduce the overall scheduling time for the whole group - as we're waiting them to be all present in the queue first and then start the full scheduling cycle for each).
 
-2. `PreFilter` validates that if we should reject `WaitingPods` in advance. When the pod belongs to a new `PodGroup` (we can check it by the `LastPodGroup` in cache), it indicates that the new PodGroup scheduling cycle has been entered. Then we will check if the last `PodGroup` meets the minAvailable, if not, we will reject it directly in advance. The purpose is 1.To avoid invalid waiting, 2.To make the pod belongs to the same `PodGroup` fail together, rather than waiting partially.
+Here we're adopting the latter approach, `PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.
```

Comment:
Are we still removing the PreFilter?

Reply:
I remove the mechanism to reject previous `PodGroup`, just left that ``PreFilter` validates that if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`. If so, the scheduling process will be interrupted directly.``

Reply:
the total number of pods where? i.e. where does this counter get increased?

Reply:
> the total number of pods where? i.e. where does this counter get increased?
I get the total number of Pods belong to the same `PodGroup` through `PodLister.List(selector)` because they all have the same label `pod-group.scheduling.sigs.k8s.io/name` @alculquicondor

Reply:
That would bypass all the caches, which may led to race conditions. Also it moves pods to the "unschedulable" queue if the number is lower than min. Then there is no way to bring them back unless some condition in the cluster changes that makes the pod "more schedulable". I know there is the flush every one minute, which I want to get rid of. But you probably don't want the plugin to have to wait 1min anyways.

Reply:
> That would bypass all the caches

The PodLister should come from snapshot, so should be racing free. It's just that the `PodLister.List(filter)` does simple filtering instead of index, so perf can be improved.

> Then there is no way to bring them back unless some condition in the cluster changes that makes the pod "more schedulable".

I'm also concerned about this. However, think about a regular coscheduling implementation - i.e., which schedules every PodGroup Pod unconditionally and rely on Preempt plugin to wait to reach minAvailable, in that case, it seems this problem also exists?

Reply:
That lister doesn't include unscheduled pods, which these are.

> However, think about a regular coscheduling implementation - i.e., which schedules every PodGroup Pod unconditionally and rely on Preempt plugin to wait to reach minAvailable, in that case, it seems this problem also exists?

That's not the case, because when a pod gets deleted, we bring back pods from the unschedulable queue back to the active queue.

Reply:
> That lister doesn't include unscheduled pods, which these are.

That's true.

> That's not the case, because when a pod gets deleted, we bring back pods from the unschedulable queue back to the active queue.

But in both implementations, no (waiting) Pods get deleted - they just get moved from activeQ to unschedulableQ/packoffQ, or the other way around.

Reply:
You said "Preempt plugin". I thought you were just referring to preemption. Are you referring to Reserve?
In that case, pods move to the plugin's cache. They are neither in active or unschedulableQ.

Reply:
oops... my bad, typo, I meant "Permit" plugins.

Reply:
No worries. So my comment still applies. We are not moving pods to unschedulable queue when we tell pods to "wait".

Reply:
True. My point was that based on current extension points, this seems to be a common limitation among all coscheduling plugins.